File: telegram-bot-sdk/node/src/commands/scrum.mjs
/**
 * commands/scrum.mjs — Scrum board Telegram commands
 *
 * Registers the following commands on the Telegraf bot:
 *  /sprint  [create|list|close|activate|status] [name]
 *  /item    [add|done|start|block|list] [sprint?] [title?]
 *  /board   [sprint?]    — Kanban-style board view
 *  /velocity [sprint?]   — Sprint velocity summary
 *
 * All commands accept the active sprint as default when no sprint ID is given.
 */

// ---------------------------------------------------------------------------
// Formatting helpers
// ---------------------------------------------------------------------------

const STATUS_EMOJI = {
  todo: '⬜',
  'in-progress': '🔵',
  done: '✅',
  blocked: '🔴',
};

const TYPE_EMOJI = {
  story: '📖',
  task: '🔧',
  bug: '🐛',
  spike: '⚡',
};

const SPRINT_STATUS_EMOJI = {
  planned: '📅',
  active: '🚀',
  closed: '🏁',
};

function fmtSprint(sprint) {
  const emoji = SPRINT_STATUS_EMOJI[sprint.status] ?? '📋';
  const closed = sprint.closedAt
    ? ` | Closed: ${sprint.closedAt.slice(0, 10)}`
    : '';
  return `${emoji} *${escMd(sprint.name)}* \`${sprint.id.slice(0, 8)}\`\nStatus: ${sprint.status}${closed}`;
}

function fmtItem(item) {
  const se = STATUS_EMOJI[item.status] ?? '•';
  const te = TYPE_EMOJI[item.type] ?? '•';
  const pts = item.points > 0 ? ` _(${item.points}pt)_` : '';
  const who = item.assignee ? ` @${escMd(item.assignee)}` : '';
  return `${se}${te} ${escMd(item.title)}${pts}${who}`;
}

function escMd(text) {
  // Escape special chars for Telegram MarkdownV2
  return String(text).replace(/[_*[\]()~`>#+=|{}.!\\-]/g, (c) => `\\${c}`);
}

// ---------------------------------------------------------------------------
// Argument parsing helpers
// ---------------------------------------------------------------------------

/**
 * Parse command text: "/sprint create My Sprint"  → { sub: 'create', rest: 'My Sprint' }
 */
function parseArgs(text) {
  const parts = (text ?? '').trim().replace(/^\/\w+\s*/, '').split(/\s+/);
  const sub = parts[0]?.toLowerCase() ?? '';
  const rest = parts.slice(1).join(' ').trim();
  return { sub, rest, parts: parts.slice(1) };
}

// ---------------------------------------------------------------------------
// ScrumCommands
// ---------------------------------------------------------------------------

export class ScrumCommands {
  #board;

  /** @param {{ board: import('../lib/scrum-board.mjs').ScrumBoard }} opts */
  constructor({ board }) {
    this.#board = board;
  }

  /**
   * Register all scrum commands with a Telegraf bot instance.
   * @param {import('telegraf').Telegraf} bot
   */
  register(bot) {
    bot.command('sprint', (ctx) => this.#handleSprint(ctx));
    bot.command('item', (ctx) => this.#handleItem(ctx));
    bot.command('board', (ctx) => this.#handleBoard(ctx));
    bot.command('velocity', (ctx) => this.#handleVelocity(ctx));
  }

  // -------------------------------------------------------------------------
  // /sprint
  // -------------------------------------------------------------------------

  async #handleSprint(ctx) {
    const { sub, rest } = parseArgs(ctx.message?.text);

    try {
      switch (sub) {
        case 'create':
          return ctx.reply(this.#sprintCreate(rest), { parse_mode: 'Markdown' });
        case 'activate':
          return ctx.reply(this.#sprintActivate(rest), { parse_mode: 'Markdown' });
        case 'close':
          return ctx.reply(this.#sprintClose(rest), { parse_mode: 'Markdown' });
        case 'list':
        case '':
          return ctx.reply(this.#sprintList(rest), { parse_mode: 'Markdown' });
        case 'status':
          return ctx.reply(this.#sprintStatus(rest), { parse_mode: 'Markdown' });
        default:
          return ctx.reply(this.#sprintHelp(), { parse_mode: 'Markdown' });
      }
    } catch (err) {
      return ctx.reply(`❌ ${err.message}`);
    }
  }

  #sprintCreate(name) {
    if (!name) return '⚠️ Usage: `/sprint create <name>`';
    const sprint = this.#board.createSprint({ name });
    return (
      `✅ Sprint created!\n\n` +
      `${fmtSprint(sprint)}\n\n` +
      `Use \`/sprint activate ${sprint.id.slice(0, 8)}\` to start it.`
    );
  }

  #sprintActivate(idOrName) {
    const sprint = this.#resolveSprint(idOrName);
    const activated = this.#board.activateSprint(sprint.id);
    return `🚀 Sprint *${escMd(activated.name)}* is now active!`;
  }

  #sprintClose(idOrName) {
    const sprint = this.#resolveSprint(idOrName);
    const closed = this.#board.closeSprint(sprint.id);
    const { completed, committed, percentage } = this.#board.getVelocity(closed.id);
    return (
      `🏁 Sprint *${escMd(closed.name)}* closed\\.\n\n` +
      `Velocity: ${completed}/${committed} pts \\(${percentage}%\\)`
    );
  }

  #sprintList(statusFilter) {
    const validStatuses = ['active', 'planned', 'closed'];
    const status = validStatuses.includes(statusFilter) ? statusFilter : undefined;
    const sprints = this.#board.listSprints({ status });

    if (sprints.length === 0) {
      return status
        ? `No ${status} sprints found. Use \`/sprint create <name>\` to start one.`
        : 'No sprints yet. Use `/sprint create <name>` to start one.';
    }

    const lines = sprints.map((s) => fmtSprint(s));
    return `*Sprints* (${sprints.length})\n\n${lines.join('\n\n')}`;
  }

  #sprintStatus(idOrName) {
    const sprint = this.#resolveSprint(idOrName);
    const { committed, completed, percentage } = this.#board.getVelocity(sprint.id);
    const items = this.#board.listItems(sprint.id);

    const statusCounts = { todo: 0, 'in-progress': 0, done: 0, blocked: 0 };
    for (const item of items) statusCounts[item.status]++;

    return (
      `${fmtSprint(sprint)}\n\n` +
      `📊 *Progress*: ${completed}/${committed} pts (${percentage}%)\n` +
      `⬜ Todo: ${statusCounts.todo}\n` +
      `🔵 In-progress: ${statusCounts['in-progress']}\n` +
      `✅ Done: ${statusCounts.done}\n` +
      `🔴 Blocked: ${statusCounts.blocked}`
    );
  }

  #sprintHelp() {
    return (
      `*Sprint Commands*\n\n` +
      `\`/sprint create <name>\` — Create a new sprint\n` +
      `\`/sprint list [status]\` — List sprints\n` +
      `\`/sprint activate <id>\` — Activate a sprint\n` +
      `\`/sprint close <id>\` — Close a sprint\n` +
      `\`/sprint status [id]\` — Sprint status`
    );
  }

  // -------------------------------------------------------------------------
  // /item
  // -------------------------------------------------------------------------

  async #handleItem(ctx) {
    const { sub, rest, parts } = parseArgs(ctx.message?.text);

    try {
      switch (sub) {
        case 'add':
          return ctx.reply(await this.#itemAdd(parts), { parse_mode: 'Markdown' });
        case 'done':
          return ctx.reply(this.#itemDone(parts[0]), { parse_mode: 'Markdown' });
        case 'start':
          return ctx.reply(this.#itemStart(parts[0]), { parse_mode: 'Markdown' });
        case 'block':
          return ctx.reply(this.#itemBlock(parts[0]), { parse_mode: 'Markdown' });
        case 'list':
        case '':
          return ctx.reply(this.#itemList(parts[0]), { parse_mode: 'Markdown' });
        default:
          return ctx.reply(this.#itemHelp(), { parse_mode: 'Markdown' });
      }
    } catch (err) {
      return ctx.reply(`❌ ${err.message}`);
    }
  }

  async #itemAdd(parts) {
    // /item add [sprint-id] <title>
    // If first part matches a sprint ID, use it; otherwise use active sprint
    let sprintId;
    let titleParts = parts;

    const firstPart = parts[0];
    const matchedSprint = this.#tryResolveSprint(firstPart);
    if (matchedSprint) {
      sprintId = matchedSprint.id;
      titleParts = parts.slice(1);
    } else {
      const active = this.#board.getActiveSprint();
      if (!active) throw new Error('No active sprint. Specify a sprint ID.');
      sprintId = active.id;
    }

    const title = titleParts.join(' ').trim();
    if (!title) throw new Error('Usage: `/item add [sprint-id] <title>`');

    const item = this.#board.addItem({ sprintId, title });
    const sprint = this.#board.getSprint(sprintId);

    return (
      `✅ Item added to *${escMd(sprint.name)}*\n\n` +
      `${fmtItem(item)}\n` +
      `ID: \`${item.id.slice(0, 8)}\``
    );
  }

  #itemDone(itemId) {
    if (!itemId) throw new Error('Usage: `/item done <item-id>`');
    const item = this.#resolveItem(itemId);
    const updated = this.#board.completeItem(item.id);
    return `✅ *${escMd(updated.title)}* marked as done!`;
  }

  #itemStart(itemId) {
    if (!itemId) throw new Error('Usage: `/item start <item-id>`');
    const item = this.#resolveItem(itemId);
    const updated = this.#board.startItem(item.id);
    return `🔵 *${escMd(updated.title)}* is now in-progress.`;
  }

  #itemBlock(itemId) {
    if (!itemId) throw new Error('Usage: `/item block <item-id>`');
    const item = this.#resolveItem(itemId);
    const updated = this.#board.blockItem(item.id);
    return `🔴 *${escMd(updated.title)}* marked as blocked.`;
  }

  #itemList(sprintIdOrEmpty) {
    let sprintId;
    if (sprintIdOrEmpty) {
      sprintId = this.#resolveSprint(sprintIdOrEmpty).id;
    } else {
      const active = this.#board.getActiveSprint();
      if (!active) throw new Error('No active sprint. Specify a sprint ID.');
      sprintId = active.id;
    }

    const items = this.#board.listItems(sprintId);
    if (items.length === 0) return 'No items in this sprint yet.';

    const sprint = this.#board.getSprint(sprintId);
    const lines = items.map((i) => `${fmtItem(i)} \`${i.id.slice(0, 8)}\``);
    return `*Items in ${escMd(sprint.name)}*\n\n${lines.join('\n')}`;
  }

  #itemHelp() {
    return (
      `*Item Commands*\n\n` +
      `\`/item add [sprint-id] <title>\` — Add item\n` +
      `\`/item list [sprint-id]\` — List items\n` +
      `\`/item start <id>\` — Start item\n` +
      `\`/item done <id>\` — Complete item\n` +
      `\`/item block <id>\` — Block item`
    );
  }

  // -------------------------------------------------------------------------
  // /board
  // -------------------------------------------------------------------------

  async #handleBoard(ctx) {
    const { rest } = parseArgs(ctx.message?.text);
    try {
      let sprintId;
      if (rest) {
        sprintId = this.#resolveSprint(rest).id;
      } else {
        const active = this.#board.getActiveSprint();
        if (!active) return ctx.reply('No active sprint. Use `/sprint activate <id>`.');
        sprintId = active.id;
      }

      const { sprint, columns, velocity } = this.#board.getBoardView(sprintId);

      const render = (items) =>
        items.length === 0
          ? '_empty_'
          : items.map((i) => `  ${fmtItem(i)}`).join('\n');

      const msg =
        `📋 *Board: ${escMd(sprint.name)}*\n\n` +
        `⬜ *Todo* (${columns.todo.length})\n${render(columns.todo)}\n\n` +
        `🔵 *In-Progress* (${columns['in-progress'].length})\n${render(columns['in-progress'])}\n\n` +
        `✅ *Done* (${columns.done.length})\n${render(columns.done)}\n\n` +
        `🔴 *Blocked* (${columns.blocked.length})\n${render(columns.blocked)}\n\n` +
        `⚡ *Velocity*: ${velocity} pts completed`;

      return ctx.reply(msg, { parse_mode: 'Markdown' });
    } catch (err) {
      return ctx.reply(`❌ ${err.message}`);
    }
  }

  // -------------------------------------------------------------------------
  // /velocity
  // -------------------------------------------------------------------------

  async #handleVelocity(ctx) {
    const { rest } = parseArgs(ctx.message?.text);
    try {
      let sprintId;
      if (rest) {
        sprintId = this.#resolveSprint(rest).id;
      } else {
        const active = this.#board.getActiveSprint();
        if (!active) return ctx.reply('No active sprint.');
        sprintId = active.id;
      }

      const sprint = this.#board.getSprint(sprintId);
      const { committed, completed, percentage } = this.#board.getVelocity(sprintId);

      const bar = buildProgressBar(percentage);
      return ctx.reply(
        `⚡ *Velocity: ${escMd(sprint.name)}*\n\n` +
          `${bar} ${percentage}%\n` +
          `Completed: ${completed} / ${committed} story points`,
        { parse_mode: 'Markdown' }
      );
    } catch (err) {
      return ctx.reply(`❌ ${err.message}`);
    }
  }

  // -------------------------------------------------------------------------
  // Resolution helpers
  // -------------------------------------------------------------------------

  #resolveSprint(idOrName) {
    const sprint = this.#tryResolveSprint(idOrName);
    if (!sprint) throw new Error(`Sprint not found: "${idOrName}"`);
    return sprint;
  }

  #tryResolveSprint(idOrName) {
    if (!idOrName) return null;
    const all = this.#board.listSprints();
    // Try exact ID match (prefix)
    const byId = all.find((s) => s.id.startsWith(idOrName));
    if (byId) return byId;
    // Try case-insensitive name match
    const byName = all.find(
      (s) => s.name.toLowerCase() === idOrName.toLowerCase()
    );
    return byName ?? null;
  }

  #resolveItem(idOrTitle) {
    // Search across all sprints
    const sprints = this.#board.listSprints();
    for (const sprint of sprints) {
      const items = this.#board.listItems(sprint.id);
      const found = items.find(
        (i) =>
          i.id.startsWith(idOrTitle) ||
          i.title.toLowerCase() === idOrTitle.toLowerCase()
      );
      if (found) return found;
    }
    throw new Error(`Item not found: "${idOrTitle}"`);
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function buildProgressBar(pct, len = 10) {
  const filled = Math.round((pct / 100) * len);
  return '█'.repeat(filled) + '░'.repeat(len - filled);
}

export default ScrumCommands;

File: telegram-bot-sdk/rust/src/commands/scrum.rs
//! Scrum-related command handlers.
//!
//! Commands:
//!  - `/sprint [create|list|activate|complete|status] [args...]`
//!  - `/standup` — generate a daily standup summary
//!  - `/velocity [sprint-id-prefix]` — sprint velocity report
//!  - `/status` — active sprint status at a glance
//!
//! All commands hold an `Arc<AppState>` so they can read/write shared state.

use std::sync::Arc;
use async_trait::async_trait;
use chrono::{Duration, Utc};
use tracing::info;

use crate::commands::Command;
use crate::error::{BotError, Result};
use crate::state::{AppState, Sprint, SprintStatus};
use crate::types::{Context, Response};

// ---------------------------------------------------------------------------
// SprintCommand — /sprint
// ---------------------------------------------------------------------------

/// Handles all `/sprint` subcommands.
pub struct SprintCommand {
    state: Arc<AppState>,
}

impl SprintCommand {
    pub fn new(state: Arc<AppState>) -> Self {
        Self { state }
    }
}

#[async_trait]
impl Command for SprintCommand {
    fn name(&self) -> &str { "sprint" }
    fn description(&self) -> &str { "Sprint management" }
    fn usage(&self) -> &str { "[create|list|activate|complete|status] [args]" }

    async fn execute(&self, ctx: &Context, args: &[&str]) -> Result<Response> {
        let sub = args.first().copied().unwrap_or("list");
        match sub {
            "create" | "new" => self.create(ctx, &args[1..]).await,
            "list" | "ls" => self.list(ctx).await,
            "activate" | "start" => self.activate(ctx, &args[1..]).await,
            "complete" | "close" | "done" => self.complete(ctx, &args[1..]).await,
            "status" | "info" => self.status(ctx, &args[1..]).await,
            "help" | _ => Ok(Response::text(self.sprint_help())),
        }
    }
}

impl SprintCommand {
    async fn create(&self, ctx: &Context, args: &[&str]) -> Result<Response> {
        if args.is_empty() {
            return Ok(Response::text(
                "Usage: /sprint create <name> [duration_days]\n\
                 Example: /sprint create Sprint-42 14",
            ));
        }

        let name = args[0];
        let days: i64 = args.get(1).and_then(|s| s.parse().ok()).unwrap_or(14);
        let now = Utc::now();
        let sprint = Sprint::new(name, now, now + Duration::days(days));
        let id_short = sprint.id.to_string()[..8].to_owned();
        let created = self.state.create_sprint(sprint).await?;

        info!("Sprint created: {} ({})", created.name, created.id);
        Ok(Response::text(format!(
            "Sprint '{}' created!\n\
             ID: {}\n\
             Duration: {} days\n\
             Start: {}\n\
             End: {}\n\n\
             Use /sprint activate {} to start it.",
            created.name,
            id_short,
            days,
            created.start_date.format("%Y-%m-%d"),
            created.end_date.format("%Y-%m-%d"),
            id_short,
        )))
    }

    async fn list(&self, _ctx: &Context) -> Result<Response> {
        let sprints = self.state.list_sprints().await;
        if sprints.is_empty() {
            return Ok(Response::text(
                "No sprints yet. Use /sprint create <name> to create one.",
            ));
        }

        let mut lines = vec!["Sprints:".to_string()];
        for s in &sprints {
            let marker = if s.status == SprintStatus::Active { " [ACTIVE]" } else { "" };
            lines.push(format!(
                "  [{}] {} — {}{}",
                &s.id.to_string()[..8],
                s.name,
                s.status,
                marker,
            ));
        }
        Ok(Response::text(lines.join("\n")))
    }

    async fn activate(&self, _ctx: &Context, args: &[&str]) -> Result<Response> {
        let prefix = args.first().copied().unwrap_or("");
        if prefix.is_empty() {
            return Ok(Response::text("Usage: /sprint activate <id-prefix>"));
        }

        let sprints = self.state.list_sprints().await;
        let sprint = sprints
            .iter()
            .find(|s| s.id.to_string().starts_with(prefix) || s.name == prefix)
            .ok_or_else(|| BotError::StateError(format!("Sprint '{}' not found", prefix)))?;

        let id = sprint.id;
        self.state.set_active_sprint(id).await?;
        info!("Sprint activated: {}", sprint.name);

        Ok(Response::text(format!(
            "Sprint '{}' is now ACTIVE!\n\
             Duration: {} days\n\
             Ends: {}",
            sprint.name,
            sprint.duration_days(),
            sprint.end_date.format("%Y-%m-%d"),
        )))
    }

    async fn complete(&self, _ctx: &Context, args: &[&str]) -> Result<Response> {
        let prefix = args.first().copied();
        let sprint = if let Some(p) = prefix {
            let sprints = self.state.list_sprints().await;
            sprints
                .into_iter()
                .find(|s| s.id.to_string().starts_with(p) || s.name == p)
                .ok_or_else(|| BotError::StateError(format!("Sprint '{}' not found", p)))?
        } else {
            self.state
                .get_active_sprint()
                .await
                .ok_or_else(|| BotError::StateError("No active sprint".to_string()))?
        };

        let id = sprint.id;
        let (committed, completed) = self.state.sprint_velocity(id).await;
        let pct = if committed > 0 { completed * 100 / committed } else { 0 };

        self.state
            .update_sprint(id, |s| {
                s.status = SprintStatus::Completed;
                s.velocity = Some(completed as f64);
            })
            .await?;

        info!("Sprint completed: {}", sprint.name);
        Ok(Response::text(format!(
            "Sprint '{}' completed!\n\
             Story points: {}/{} ({}%)\n\
             Velocity: {} pts",
            sprint.name, completed, committed, pct, completed
        )))
    }

    async fn status(&self, _ctx: &Context, args: &[&str]) -> Result<Response> {
        let prefix = args.first().copied();
        let sprint = if let Some(p) = prefix {
            let sprints = self.state.list_sprints().await;
            sprints
                .into_iter()
                .find(|s| s.id.to_string().starts_with(p) || s.name == p)
                .ok_or_else(|| BotError::StateError(format!("Sprint '{}' not found", p)))?
        } else {
            self.state
                .get_active_sprint()
                .await
                .ok_or_else(|| BotError::StateError("No active sprint".to_string()))?
        };

        let issues = self.state.list_issues_by_sprint(sprint.id).await;
        let done = issues.iter().filter(|i| i.is_complete()).count();
        let (committed, completed) = self.state.sprint_velocity(sprint.id).await;
        let pct = if committed > 0 { completed * 100 / committed } else { 0 };

        Ok(Response::text(format!(
            "Sprint: {}\n\
             Status: {}\n\
             Days remaining: {}\n\
             Issues: {}/{} done\n\
             Story points: {}/{} ({}%)",
            sprint.name,
            sprint.status,
            sprint.days_remaining(),
            done,
            issues.len(),
            completed,
            committed,
            pct,
        )))
    }

    fn sprint_help(&self) -> String {
        "Sprint Commands:\n\
         /sprint create <name> [days] — Create a new sprint\n\
         /sprint list — List all sprints\n\
         /sprint activate <id> — Activate a sprint\n\
         /sprint complete [id] — Complete a sprint\n\
         /sprint status [id] — Sprint details"
            .to_string()
    }
}

// ---------------------------------------------------------------------------
// StandupCommand — /standup
// ---------------------------------------------------------------------------

/// Generates a daily standup digest: what was done, what's in progress,
/// what's blocked, for the active sprint.
pub struct StandupCommand {
    state: Arc<AppState>,
}

impl StandupCommand {
    pub fn new(state: Arc<AppState>) -> Self {
        Self { state }
    }
}

#[async_trait]
impl Command for StandupCommand {
    fn name(&self) -> &str { "standup" }
    fn description(&self) -> &str { "Generate daily standup report for the active sprint" }

    async fn execute(&self, _ctx: &Context, _args: &[&str]) -> Result<Response> {
        let sprint = self.state
            .get_active_sprint()
            .await
            .ok_or_else(|| BotError::StateError("No active sprint".to_string()))?;

        let issues = self.state.list_issues_by_sprint(sprint.id).await;

        use crate::state::IssueStatus;
        let done: Vec<_> = issues
            .iter()
            .filter(|i| i.status == IssueStatus::Done)
            .collect();
        let in_progress: Vec<_> = issues
            .iter()
            .filter(|i| i.status == IssueStatus::InProgress)
            .collect();
        let blocked: Vec<_> = issues
            .iter()
            .filter(|i| i.status == IssueStatus::InReview)
            .collect();
        let todo: Vec<_> = issues
            .iter()
            .filter(|i| i.status == IssueStatus::Todo || i.status == IssueStatus::Backlog)
            .collect();

        let fmt_issues = |items: &[&crate::state::Issue]| -> String {
            if items.is_empty() {
                return "  (none)".to_string();
            }
            items
                .iter()
                .map(|i| {
                    let assignee = i
                        .assignee
                        .as_deref()
                        .map(|a| format!(" (@{})", a))
                        .unwrap_or_default();
                    let pts = i.story_points.map(|p| format!(" [{}pt]", p)).unwrap_or_default();
                    format!("  - {}{}{}", i.title, pts, assignee)
                })
                .collect::<Vec<_>>()
                .join("\n")
        };

        let (committed, completed) = self.state.sprint_velocity(sprint.id).await;
        let pct = if committed > 0 { completed * 100 / committed } else { 0 };

        Ok(Response::text(format!(
            "Daily Standup — {}\n\
             Date: {}\n\
             Days remaining: {}\n\
             Velocity: {}/{} pts ({}%)\n\n\
             DONE:\n{}\n\n\
             IN PROGRESS:\n{}\n\n\
             TODO:\n{}\n\n\
             BLOCKED:\n{}",
            sprint.name,
            Utc::now().format("%Y-%m-%d"),
            sprint.days_remaining(),
            completed, committed, pct,
            fmt_issues(&done),
            fmt_issues(&in_progress),
            fmt_issues(&todo),
            fmt_issues(&blocked),
        )))
    }
}

// ---------------------------------------------------------------------------
// VelocityCommand — /velocity
// ---------------------------------------------------------------------------

pub struct VelocityCommand {
    state: Arc<AppState>,
}

impl VelocityCommand {
    pub fn new(state: Arc<AppState>) -> Self {
        Self { state }
    }
}

#[async_trait]
impl Command for VelocityCommand {
    fn name(&self) -> &str { "velocity" }
    fn description(&self) -> &str { "Show sprint velocity metrics" }
    fn usage(&self) -> &str { "[sprint-id]" }

    async fn execute(&self, _ctx: &Context, args: &[&str]) -> Result<Response> {
        let sprint = if let Some(prefix) = args.first().copied() {
            let sprints = self.state.list_sprints().await;
            sprints
                .into_iter()
                .find(|s| s.id.to_string().starts_with(prefix) || s.name == prefix)
                .ok_or_else(|| BotError::StateError(format!("Sprint '{}' not found", prefix)))?
        } else {
            self.state
                .get_active_sprint()
                .await
                .ok_or_else(|| BotError::StateError("No active sprint".to_string()))?
        };

        let (committed, completed) = self.state.sprint_velocity(sprint.id).await;
        let pct = if committed > 0 { completed * 100 / committed } else { 0u32 };
        let bar = progress_bar(pct as usize, 20);

        Ok(Response::text(format!(
            "Velocity — {}\n\
             {} {}%\n\
             Completed: {} / {} story points\n\
             Status: {}",
            sprint.name,
            bar,
            pct,
            completed,
            committed,
            sprint.status,
        )))
    }
}

// ---------------------------------------------------------------------------
// ScrumDispatcher — convenience facade used by bot.rs
// ---------------------------------------------------------------------------

/// Thin dispatcher that instantiates the right scrum command and executes it.
/// Used by `bot.rs` to avoid importing every command struct individually.
pub struct ScrumDispatcher {
    state: Arc<AppState>,
}

impl ScrumDispatcher {
    pub fn new(state: Arc<AppState>) -> Self {
        Self { state }
    }

    pub async fn dispatch(&self, ctx: &Context, args: &[&str]) -> Result<Response> {
        SprintCommand::new(self.state.clone()).execute(ctx, args).await
    }

    pub async fn standup(&self, ctx: &Context) -> Result<Response> {
        StandupCommand::new(self.state.clone()).execute(ctx, &[]).await
    }

    pub async fn velocity(&self, ctx: &Context) -> Result<Response> {
        VelocityCommand::new(self.state.clone()).execute(ctx, &[]).await
    }

    pub async fn status(&self, ctx: &Context) -> Result<Response> {
        SprintCommand::new(self.state.clone()).execute(ctx, &["status"]).await
    }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn progress_bar(pct: usize, width: usize) -> String {
    let filled = (pct.min(100) * width) / 100;
    let empty = width.saturating_sub(filled);
    format!("[{}{}]", "#".repeat(filled), ".".repeat(empty))
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::BotConfig;
    use crate::types::Message;

    fn make_state() -> Arc<AppState> {
        AppState::new(BotConfig {
            token: "test".into(),
            ..Default::default()
        })
    }

    fn ctx() -> Context {
        Context::new(Message::new(1, 1, "/sprint"))
    }

    #[tokio::test]
    async fn sprint_create_and_list() {
        let state = make_state();
        let cmd = SprintCommand::new(state.clone());

        let resp = cmd.execute(&ctx(), &["create", "Alpha", "7"]).await.unwrap();
        assert!(resp.as_str().contains("Alpha"));

        let list = cmd.execute(&ctx(), &["list"]).await.unwrap();
        assert!(list.as_str().contains("Alpha"));
    }

    #[tokio::test]
    async fn standup_no_active_sprint_error() {
        let state = make_state();
        let cmd = StandupCommand::new(state);
        let result = cmd.execute(&ctx(), &[]).await;
        assert!(result.is_err());
    }

    #[test]
    fn progress_bar_100() {
        let bar = progress_bar(100, 10);
        assert_eq!(bar, "[##########]");
    }

    #[test]
    fn progress_bar_0() {
        let bar = progress_bar(0, 10);
        assert_eq!(bar, "[..........]");
    }
}
